pkg/serviceaccounttoken/secret_cleaner.go

```diff
@@ -0,0 +1,182 @@
+	secretsQueue, err := loadSecretQueue(ctx, secretsCache)
+	if err != nil {
+		return err
+	}
+
+	startTime := time.Now()
+
+	logrus.Infof("Starting ServiceAccountSecretCleaner with %v secrets", secretsQueue.List.Len())
+	ticker := time.NewTicker(cleanCycleDelay)
+
+	secrets := client.Secrets(impersonationNamespace)
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				logrus.Info("terminating service account secret cleaner")
+				return
+			case <-ticker.C:
+				if err := CleanServiceAccountSecrets(ctx, secrets, serviceAccounts, secretsQueue.dequeue(cleaningBatchSize)); err != nil {
+					logrus.Error(err)
+				}
+				if l := secretsQueue.List.Len(); l > 0 {
+					logrus.Infof("ServiceAccountSecretCleaner has %v secrets remaining", l)
+				} else {
+					logrus.Infof("ServiceAccountSecretCleaner has no secrets remaining - terminating at %v", time.Since(startTime))
+					return
+				}
+				// This ensures that no matter how long the cleaning takes,
+				// we'll alwas keep the same cycle time.
+				ticker.Reset(cleanCycleTime)
+			}
+		}
+	}()
+
+	return nil
+}
+
+// CleanServiceAccountSecrets removes unused Secrets for ServiceAccountTokens from a
+// namespace in batches.
+func CleanServiceAccountSecrets(ctx context.Context, secrets clientv1.SecretInterface, serviceAccounts clientv1.ServiceAccountInterface, secretsToDelete []*corev1.Secret) error {
+	var deletionErr error
+	var deletedCount int64
+	var toBeDeleted []*corev1.Secret
+
+	for i := range secretsToDelete {
+		secret := secretsToDelete[i]
+		serviceAccount, err := findServiceAccountForSecret(ctx, secret, serviceAccounts)
+		if err != nil {
+			deletionErr = errors.Join(deletionErr, err)
+			continue
+		}
+
+		// If have a ServiceAccount for this secret, then we can leave it in
+		// place.
+		if serviceAccount != nil {
+			continue
+		}
+
+		toBeDeleted = append(toBeDeleted, secret)
+	}
+
+	for _, secret := range toBeDeleted {
+		logrus.Debugf("Deleting ServiceAccount Secret %s", logKeyFromObject(secret))
+		if err := secrets.Delete(ctx, secret.Name, metav1.DeleteOptions{}); err != nil {
+			deletionErr = errors.Join(deletionErr, err)
+		}
+		deletedCount++
+	}
+
+	if deletedCount > 0 {
+		logrus.Infof("SecretCleaner deleted %v secrets", deletedCount)
+	}
+
+	return deletionErr
+}
+
+func findServiceAccountForSecret(ctx context.Context, secret *corev1.Secret, serviceAccounts clientv1.ServiceAccountInterface) (*corev1.ServiceAccount, error) {
+	serviceAccountList, err := serviceAccounts.List(ctx, metav1.ListOptions{LabelSelector: labels.SelectorFromSet(map[string]string{
+		"authz.cluster.cattle.io/impersonator": "true",
+	}).String()})
+	if err != nil {
+		return nil, fmt.Errorf("failed to list service accounts when cleaning: %w", err)
+	}
+	secretAnnotation := secret.Namespace + "/" + secret.Name
+	for _, sa := range serviceAccountList.Items {
+		if sa.Annotations != nil && sa.Annotations[ServiceAccountSecretRefAnnotation] == secretAnnotation {
+			return &sa, nil
+		}
+	}
```

Comment:
Hmm, I not sure if I like iterating through all ServiceAccounts for all Secrets.

Could we perhaps add an Indexer to the cache, or build a lookup map instead?